Instagram Gallery: On WPCOM, Bypass the WPCOM API

Copons opened the pull request:
Requires D42946-code

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* On WordPress.com, avoid unnecessarily calling the WPCOM REST API.

Reasons:
- `wpcom_json_api_request_as_user` and `wpcom_json_api_request_as_blog` currently don't work as expected on WPCOM (see https://github.com/Automattic/jetpack/issues/15468).
- There's no real need to hit the WPCOM REST API when on WPCOM.

To avoid exposing lots of internal WPCOM and Keyring logic in Jetpack, I've externalized a bit of the Instagram WPCOM endpoint into an helper class, easier to use in Jetpack.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Jetpack setup: apply D42946-code and sandbox the API.
WPCOM setup: apply D43015-code, D42946-code, sandbox the API and a test site.

Smoke test the Instagram Gallery block: connect to Instagram, embed a gallery, disconnect from Instagram, and whatever else you can think of.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

Copons committed:
On WPCOM, bypass the WPCOM API

scruffian commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-instagram-gallery.php:
```diff
@@ -184,8 +200,26 @@ public function delete_instagram_access_token( $request ) {
+	 */
+	private function is_wpcom() {
+		if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+			require_lib( 'instagram-gallery-helper' );
```
This looks like a side effect because of the name of the function. I wonder if we can pick a different name, maybe `handle_wpcom`?

Copons replied:
I agree that it definitely smells.
I kinda went for this anyway because it saved a line per each function.

I guess another way could be to have `$is_wpcom` as a class property, set it in the constructor, and require the lib in the constructor as well.

scruffian replied:
Oh yeah that's a nice idea

scruffian commented:
This looks good and works as expected. I'm not very keen on the multiple `is_wpcom` calls everywhere. I wonder if it would be possible to extend this class with another which would be used on WPCOM. What do you think?

jeherve reviewed (dismissed):
This looks ok to merge, although like you already discussed in the comments, the additional `require_lib` surprised me a bit.

Copons commented:
Sorry folks, got back around to this a bit too late, but I'd like to clean up that `is_wpcom` code smell first!

Copons committed:
Simplify WPCOM setup, and fix SSR gallery on WPCOM

Copons commented:
@scruffian @jeherve In 8c73852 I've moved the WPCOM setup to the endpoint constructor, using a `$this->is_wpcom` property instead.
I've also replaced `require_lib` with `jetpack_require_lib` (which I think it's what I was supposed to do all along).

Most importantly, my previous take on the `get_gallery` part was very incorrect.
I added the WPCOM helper in the endpoint, forgetting that most of the logic happened in the JP helper, in order to be freely used by the block's SSR too!
This meant that, on WPCOM:
- The block couldn't render on the front end, throwing an error instead.
- The gallery wasn't cached!

I think if you have a look at the [JP helper](https://github.com/Automattic/jetpack/blob/8c738521877e3bc05da836d101c405af383b5952/_inc/lib/class-jetpack-instagram-gallery-helper.php#L40-L52) (I've highlighted the new WPCOM only part) it should be clear what goes on.
Main difference is that the endpoint returns an encoded response, while the WPCOM helper returns a normal array, therefore the need to encode it when caching it.

jeherve reviewed (approved):
This is nice, tests well too. 🚢

Copons commented:
r207095-wpcom
